src/ecdsapod.rs

```diff
@@ -0,0 +1,546 @@
+
+        Ok(())
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct EcdsaPod {
+    params: Params,
+    id: PodId,
+    msg: Secp256K1Scalar,
+    pk: ECDSAPublicKey<Secp256K1>,
+    proof: Proof,
+    vds_hash: Hash,
+}
+impl middleware::RecursivePod for EcdsaPod {
+    fn verifier_data(&self) -> VerifierOnlyCircuitData<C, D> {
+        let (_, circuit_data) = &*STANDARD_ECDSA_POD_DATA;
+        circuit_data.verifier_data().verifier_only
+    }
+    fn proof(&self) -> Proof {
+        self.proof.clone()
+    }
+    fn vds_root(&self) -> Hash {
+        self.vds_hash
+    }
+}
+
+static STANDARD_ECDSA_POD_DATA: LazyLock<(EcdsaPodVerifyTarget, CircuitData<F, C, D>)> =
+    LazyLock::new(|| build().expect("successful build"));
+
+fn build() -> Result<(EcdsaPodVerifyTarget, CircuitData<F, C, D>)> {
+    let params = &*pod2::backends::plonky2::DEFAULT_PARAMS;
+    let config = CircuitConfig::standard_recursion_config();
+    let mut builder = CircuitBuilder::<F, D>::new(config);
+    let ecdsa_pod_verify_target = EcdsaPodVerifyTarget::add_targets(&mut builder, params)?;
+    let rec_circuit_data = &*pod2::backends::plonky2::STANDARD_REC_MAIN_POD_CIRCUIT_DATA;
+    pod2::backends::plonky2::recursion::pad_circuit(&mut builder, &rec_circuit_data.common);
+
+    let data = timed!("EcdsaPod build", builder.build::<C>());
+    assert_eq!(rec_circuit_data.common, data.common);
+    Ok((ecdsa_pod_verify_target, data))
+}
+
+impl EcdsaPod {
+    fn _prove(
+        params: &Params,
+        vds_root: Hash,
+        msg: Secp256K1Scalar,
+        pk: ECDSAPublicKey<Secp256K1>,
+        signature: ECDSASignature<Secp256K1>,
+    ) -> Result<EcdsaPod> {
+        // 1. prove the EcdsaVerify circuit
+        let (ecdsa_verify_target, ecdsa_circuit_data) = &*ECDSA_VERIFY_DATA;
+        let mut pw = PartialWitness::<F>::new();
+        ecdsa_verify_target.set_targets(&mut pw, msg, pk, signature)?;
+
+        let ecdsa_verify_proof = timed!(
+            "prove the ecdsa signature verification (EcdsaVerifyTarget)",
+            ecdsa_circuit_data.prove(pw)?
+        );
+
+        // 2. verify the ecdsa_verify proof in a EcdsaPodVerifyTarget circuit
+
+        let (ecdsa_pod_target, circuit_data) = &*STANDARD_ECDSA_POD_DATA;
+
+        let statements = pub_self_statements(msg, pk)
+            .into_iter()
+            .map(mainpod::Statement::from)
+            .collect_vec();
+        let id: PodId = PodId(calculate_id(&statements, params));
+
+        // set targets
+        let input = EcdsaPodVerifyInput {
+            vds_root,
+            id,
+            proof: ecdsa_verify_proof,
+        };
+        let mut pw = PartialWitness::<F>::new();
+        ecdsa_pod_target.set_targets(&mut pw, &input)?;
+        let proof_with_pis = timed!(
+            "prove the ecdsa-verification proof verification (EcdsaPod proof)",
+            circuit_data.prove(pw)?
+        );
+
+        #[cfg(test)] // sanity check
```

Comment:
This sanity check is really nice, we catch issues as early as possible.  Compared to generating the proofs, this should be really quick right?  Maybe we can do this check even without `test`?

Also, you could also verify the first proof `ecdsa_verify_proof`!

Reply:
Regarding doing it without `test`, it might add a bit of unnecessary delay when using it then (just checked, it's around `11ms`).

Regarding adding the check for the `ecdsa_verify_proof`, cool, added.

Reply:
11ms is negligible against 40s.  I would still remove the `test` conditional compilation.  But it's not a hard request from my side.

Reply:
cool, updated it to run also in non-tests: https://github.com/0xPARC/introduction-pods/compare/600f0d60a8e5c0509274e320706be3070ef4901a..95473974f4f91416e720514619555988fa6f7b61